include/common/nmea.h

```diff
@@ -0,0 +1,205 @@
+    PPSFIX = 3,
+    RTK = 4,
+    FRTK = 5
+};
+
+struct TimeStamp
+{
+    int hour{}, minute{}, second{}, millisecond{};
+
+    TimeStamp() = default;
+
+    TimeStamp(const std::string &timeString)
+    {
+        if (timeString.empty())
+            throw GPSError("Empty string when reading time");
+        if (timeString.length() < 9)
+            throw GPSError("Time information parse error");
+        hour = stoi(timeString.substr(0, 2));
+        minute = stoi(timeString.substr(2, 2));
+        second = stoi(timeString.substr(4, 2));
+        millisecond = stoi(timeString.substr(7, 2));
+    }
+
+    std::string str() const
+    {
+        std::stringstream ss;
+        ss << std::setw(2) << std::setfill('0') << hour << ":"
+           << std::setw(2) << std::setfill('0') << minute << ":";
+        if (second < 10) {
+            ss << '0';
+        }
+        ss << static_cast<float>(second) + static_cast<float>(millisecond) / 1000.f;
+        return ss.str();
+    }
+};
+
+struct GPSData
+{
+    BoBRobotics::MapCoordinate::GPSCoordinate coordinate; // Latitude and longitude coordinate
+    units::length::meter_t altitude;                      // Altitude above ellipsoid
+    int numberOfSatellites;                               // Currently observed number of satelites
+    double horizontalDilution;                            // horizontal dilution - lower value is better
+    GPSQuality gpsQuality;                                // GPS quality indicator
+    TimeStamp time;                                       // time of measurement
+};
+
+namespace NMEA {
+
+template<size_t N>
+bool
+splitString(const std::string &line,
+            std::array<std::string, N> &fields,
+            const char *sentenceId)
+{
+    std::string field;
+    std::istringstream lineStream{ line };
+
+    // The first field indicates what kind of message we have
+    BOB_ASSERT(getline(lineStream, field, ','));
+    if (field != sentenceId) {
+        return false;
+    }
+
+    // Extract the remaining fields
+    size_t count = 0;
+    for (; getline(lineStream, field, ',') && count < N; count++) {
+        fields[count] = std::move(field);
+    }
+    if (count != N) {
+        throw GPSError{ "Bad number of fields in NMEA string" };
+    }
+
+    return true;
+}
+
+inline bool
+parse(const std::string &toParse, GPSData &data)
+{
+    using namespace std;
+    using meter_t = units::length::meter_t;
+    using degree_t = units::angle::degree_t;
+    using arcminute_t = units::angle::arcminute_t;
+
+    if (toParse.empty()) {
+        return false;
+    }
+
+    degree_t latitude, longitude;
+    arcminute_t latitudeMinutes, longitudeMinutes;
+    char latDirection, longDirection;
+    meter_t altitude;
+    double horizontalDilution;
+    int numberOfSatellites;
+    int gpsQualityIndicator;
+    GPSQuality qualityOfGps;
+
+    // Let's reuse our data structure so we don't have to reallocate with every call
+    static array<string, 9> elements;
```

Comment:
I think this will be entirely stack allocated and this makes it very un thread safe so I don't think this is a good idea

Reply:
Hmm yeah. I was thinking that this way we could avoid heap allocations but I suppose that's not really likely to be much of a problem (we're only reading at 5Hz anyway).

Reply:
I think at 5Hz, a few string allocations'll be alright (I was clearly being dense before - I see what you were trying to do now) but, if it does become slow, turning this into an ``NMEAParser`` class or something would be nicer.